### PR TITLE
Decoration improvements

### PR DESCRIPTION
**Pull request by Hixie**
Documentation changes, and, in particular, the ability to lerp from
and to any decoration by lerping via null.

### CODE REVIEW

File: packages/flutter/lib/src/painting/decoration.dart [hunk 1]
```diff
@@ -44,27 +44,72 @@ abstract class Decoration extends Diagnosticable {
...
 
-  /// Linearly interpolates from [begin] to [end].
+  /// Linearly interpolates from `begin` to `end`.
   ///
-  /// This defers to [end]'s [lerpTo] function if [end] is not null,
-  /// otherwise it uses [begin]'s [lerpFrom] function.
+  /// This defers to `end`'s [lerpTo] function if `end` is not null. If `end` is
+  /// null or if its [lerpTo] returns null, it uses `begin`'s [lerpFrom]
+  /// function instead. If both return null, it attempts to lerp from `begin` to
+  /// null before `t=0.5`, then null to `end` after `t=0.5`.
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/decoration.dart]: This almost sounds like it attempts to lerp in both ways (which wouldn't make any sense).  Maybe rephrase it "If both return null, it attempts to lerp from `begin` to null if `t<0.5` or null to `end` if `t>=0.5`"

File: packages/flutter/lib/src/painting/decoration.dart [hunk 2]
```diff
@@ -44,27 +44,72 @@ abstract class Decoration extends Diagnosticable {
...
-      return begin.lerpTo(end, t);
-    return null;
+      result = end.lerpFrom(begin, t);
+    if (result == null && begin != null)
+      result = begin.lerpTo(end, t);
+    if (result == null && begin != null && end != null) {
+      if (t < 0.5) {
+        result = begin.lerpTo(null, t * 2.0);
+      } else {
+        result = end.lerpFrom(null, (t - 0.5) * 2.0);
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/decoration.dart]: Why `(t - 0.5) * 2.0`?  Is it intended that with `t==0.5`, this will be zero?

File: packages/flutter/lib/src/painting/decoration.dart [hunk 3]
```diff
@@ -44,27 +44,72 @@ abstract class Decoration extends Diagnosticable {
...
+  /// cannot interpolate from `a`. In that case, [lerp] will try `a`'s [lerpTo]
+  /// method instead.
+  ///
+  /// Supporting interpolating from null is recommended as the [Decoration.lerp]
+  /// method uses this as a fallback when two classes can't interpolate between
+  /// each other.
+  ///
+  /// Instead of calling this directly, use [Decoration.lerp].
+  @protected
+  Decoration lerpFrom(Decoration a, double t) => null;
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/decoration.dart]: Can you document `t` in these?

Comment (Hixie) [reply packages/flutter/lib/src/painting/decoration.dart]: This affects a ton of code so filed https://github.com/flutter/flutter/issues/12377.

File: packages/flutter/lib/src/painting/decoration.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/decoration.dart]: Yeah I guess I was viewing this in terms of how the method is usually used, with t varying over time.

FIxed.

File: packages/flutter/lib/src/painting/decoration.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/decoration.dart]: Yes, unless I'm thinking about it wrong.

It's intended to go map 0..0.5 to 0..1 for the lerpTo null, and 0.5..1.0 to 0.0..1.0 for the lerpFrom null. That way the animation from 0 to 1 overall goes from begin to null to end.

**Comment by Hixie** [comment]
PTAL. I'll submit a separate PR for the `t` documentation momentarily.

**Review by tvolkert — APPROVED** [review]
LGTM
